### PR TITLE
[ENH] automatically determine which dwis can/should be merged

### PR DESCRIPTION
**Pull request by mattcieslak**
Unlike fmriprep, fieldmaps are dealt with at the very beginning of the pipeline to figure out which scans are in the same "warped space".  This PR should be an entry point for discussing how to group scans. 

This particular setup works for 
 - HCP
 - HCP lifespan
 - ABCD

### CODE REVIEW

**Comment by dPys** [comment]
I like this @mattcieslak . Curious how much the sdc performance improves with this setup when running on the HCP data that you mentioned? 

Since clear standards for combining dwi's across runs don't really exist yet in the field (and particularly for optimizing sdc accordingly), I agree that this is an important conversation to be having early on.

Another thing that might help here is to make more efficient use of aggregated metadata across runs with a volume legend to keep track of which b0s/dwis go with which sdc method/files and acq. params. Thoughts?

@dPys

**Comment by oesteban** [comment]
@mattcieslak, could you point me to the exact spot where the heuristic of which dwis can/should be merged is implemented?

**Comment by mattcieslak** [comment]
It's here: https://github.com/nipreps/dmriprep/blob/b9e4cc6ef19b5c884377c15d55dde6812a79fcf7/dmriprep/workflows/base.py#L765

File: dmriprep/workflows/base.py [hunk 1]
```diff
@@ -331,176 +437,620 @@ def init_single_subject_wf(
...
+            'dwi_series_pedir': pe_dir})
+
+    if prefer_dedicated_fmaps:
+        rationale.append('\nStrictly using data from fmaps/ for SDC')
+        LOGGER.info('\n'.join(rationale))
+        return complete_groups
+
+    # Ignoring dedicated fieldmaps, define warped spaces by PEDir
+    match_header = 'No opportunities for bidirectional PEPolar DWI series were found.'
+    group_keys = set(list(pedir_groups.keys()))
```
Comment (oesteban) [inline dmriprep/workflows/base.py]:
```suggestion
    group_keys = set(pedir_groups.keys())
```
set will consume the generator directly

File: dmriprep/workflows/base.py [hunk 2]
```diff
@@ -331,176 +437,620 @@ def init_single_subject_wf(
...
+            match = matches[0]
+            match_scan_list = pedir_groups[match]
+            if using_eddy:
+                matches_rationale.append(
+                    '  * A single group containing %s/%s PE dir groups was sent to TOPUP' % (
+                        key, match))
+                final_groups.append(
+                    {'dwi_series': scan_list,
+                     'fieldmap_info': {'suffix': 'rpe_series', 'rpe_series': match_scan_list,},
+                     'dwi_series_pedir': pe_dir})
```
Comment (oesteban) [inline dmriprep/workflows/base.py]: This should be handled by the downstream workflow. I'm sure that, given the pe dir groups split the workflow could combine them for TOPUP.

File: dmriprep/workflows/base.py [hunk 3]
```diff
@@ -331,176 +437,620 @@ def init_single_subject_wf(
...
+        LOGGER.info('\n'.join(rationale))
+        return complete_groups
+
+    # Ignoring dedicated fieldmaps, define warped spaces by PEDir
+    match_header = 'No opportunities for bidirectional PEPolar DWI series were found.'
+    group_keys = set(list(pedir_groups.keys()))
+    final_groups = []
+    matches_rationale = []
+    while len(group_keys):
+        key = group_keys.pop()
```
Comment (oesteban) [inline dmriprep/workflows/base.py]: Given that we only have three (two for humans) possible PE directions, isn't it more explicit just going over them? This filtering seems a bit prone to errors.

**Comment by mattcieslak** [comment]
Not sure how I missed all these suggestions. I will take a look and write some responses. This might also make a good topic for discussion at the code sprint.

**Comment by arokem** [comment]
Any chance we can find time to tackle this tomorrow?

**Comment by mattcieslak** [comment]
I'm going to open a new PR on this.

**Pull request closed without merging**